File: zadanie/python/ellipse_fit_evaluation/ellipse_fit_evaluation.py
import csv

import cv2
import numpy as np


def evaluate_ellipse_fit(image_filename, fit_ellipse, csv_filepath='ground_truths.csv'):
    """
    Evaluate ellipse fit by comparing its parameters to the ground truth ellipse's parameters from the CSV file.
    :param image_filename: filename of the image
    :param fit_ellipse: fitted ellipse's parameters
    :param csv_filepath: filepath to the CSV file with ground truth ellipses' parameters
    :return: fitted ellipse's fit score
    """
    gt_ellipse = __get_gt_ellipse_from_csv(image_filename, csv_filepath)

    if gt_ellipse:
        if fit_ellipse:
            return __get_ellipse_fit_score(fit_ellipse, gt_ellipse)
        else:
            return 0.0
    else:
        if fit_ellipse:
            return 0.0
        else:
            return 1.0


def __get_gt_ellipse_from_csv(image_filename, csv_filepath):
    """
    Return the ground truth ellipse's parameters from the CSV file.
    :param image_filename: filename of the image
    :param csv_filepath: filepath to the CSV file with ground truth ellipses' parameters
    :return: ground truth ellipse's parameters
    """
    with open(csv_filepath, newline='') as csvfile:
        reader = csv.DictReader(csvfile)
        for row in reader:
            if row['filename'] == image_filename:
                if row['gt_ellipse_center_x'] == '':
                    gt_ellipse = None
                else:
                    gt_ellipse = {'center': (float(row['gt_ellipse_center_x']), float(row['gt_ellipse_center_y'])),
                                  'axes': (float(row['gt_ellipse_majoraxis']), float(row['gt_ellipse_minoraxis'])),
                                  'angle': int(row['gt_ellipse_angle']),
                                  'image_width': int(row['image_width']),
                                  'image_height': int(row['image_height'])}
                return gt_ellipse
        else:
            raise ValueError("Filename not found in the CSV file.")


def __get_ellipse_fit_score(fit_ellipse, gt_ellipse):
    """
    Calculate ellipse's fit score based on fitted and ground truth ellipses' parameters.
    :param fit_ellipse: fitted ellipse's parameters
    :param gt_ellipse: ground truth ellipse's parameters
    :return: fitted ellipse's fit score
    """
    fit_ellipse_image = __draw_ellipse(fit_ellipse, (gt_ellipse['image_width'], gt_ellipse['image_width']))
    gt_ellipse_image = __draw_ellipse(gt_ellipse, (gt_ellipse['image_width'], gt_ellipse['image_width']))

    return __evaluate_overlap(gt_ellipse_image, fit_ellipse_image)


def __draw_ellipse(ellipse, image_shape):
    """
    Creates a binary image containing an ellipse.
    :param ellipse: dict with 'center', 'axes' and 'angle' keys
    :param image_shape: (int, int) == (width, height) the dimensions of the image (NumPy shape)
    :return: binary image containing an ellipse
    """
    ellipse_image = np.zeros(image_shape, np.uint8)
    center = (int(np.around(ellipse['center'][0])), int(np.around(ellipse['center'][1])))
    axes = (int(np.around(ellipse['axes'][0])), int(np.around(ellipse['axes'][1])))
    angle = ellipse['angle']

    cv2.ellipse(ellipse_image, center, axes, angle, 0, 360, 1, 1)
    cv2.ellipse(ellipse_image, center, axes, angle, 0, 360, 1, cv2.FILLED)

    return ellipse_image


def __evaluate_overlap(gt_ellipse_image, fit_ellipse_image):
    """
    Calculate overlap of two binary images of ellipses.
    :param gt_ellipse_image: (numpy.ndarray) binary image with the ground truth ellipse
    :param fit_ellipse_image: (numpy.ndarray) binary image with the fitted ellipse
    :return: relative overlap of two binary images of ellipses
    """
    gt_sum = np.sum(gt_ellipse_image)
    fit_sum = np.sum(fit_ellipse_image)

    overlap_image = gt_ellipse_image & fit_ellipse_image
    overlap_sum = np.sum(overlap_image)

    return overlap_sum / max(fit_sum, gt_sum)

File: zadanie/csharp/EllipseFitEvaluation/EllipseFitEvaluation/EllipseFitEvaluation.cs
using System;
using System.IO;
using OpenCvSharp;

namespace EllipseFitEvaluation
{
    public class EllipseFitEvaluation
    {
        /// <summary>
        /// Evaluate ellipse fit by comparing its parameters to the ground truth ellipse's parameters from the CSV file.
        /// </summary>
        /// <param name="imageFilename">Filename of the image</param>
        /// <param name="fitEllipse">Fitted ellipse's parameters</param>
        /// <param name="csvFilepath">Filepath to the CSV file with ground truth ellipses' parameters</param>
        /// <returns>Fitted ellipse's fit score</returns>
        public static float EvaluateEllipseFit(string imageFilename, Ellipse fitEllipse, string csvFilepath)
        {
            GtEllipse gtEllipse = GetGtEllipseFromCsv(imageFilename, csvFilepath);

            if (gtEllipse != null)
            {
                if (fitEllipse != null)
                {
                    return GetEllipseFitScore(fitEllipse, gtEllipse);
                }
                else
                {
                    return 0.0f;
                }
            }
            else
            {
                if (fitEllipse != null)
                {
                    return 0.0f;
                }
                else
                {
                    return 1.0f;
                }
            }
        }

        /// <summary>
        /// Return the ground truth ellipse's parameters from the CSV file.
        /// </summary>
        /// <param name="imageFilename">Filename of the image</param>
        /// <param name="csvFilepath">Filepath to the CSV file with ground truth ellipses' parameters</param>
        /// <returns>Ground truth ellipse's parameters</returns>
        private static GtEllipse GetGtEllipseFromCsv(string imageFilename, string csvFilepath)
        {
            using (StreamReader sr = new StreamReader(csvFilepath))
            {
                GtEllipse gtEllipse;
                string line;
                sr.ReadLine();  // skip the header
                while ((line = sr.ReadLine()) != null)
                {
                    string[] columns = line.Split(',');
                    if (string.Compare(columns[0], imageFilename) == 0)
                    {
                        if (string.IsNullOrEmpty(columns[1]))
                        {
                            gtEllipse = null;
                        }
                        else
                        {
                            Point2f center = new Point2f(Convert.ToSingle(columns[1]), Convert.ToSingle(columns[2]));
                            Size2f axes = new Size2f(Convert.ToSingle(columns[3]), Convert.ToSingle(columns[4]));
                            float angle = Convert.ToSingle(columns[5]);
                            Size imageSize = new Size(Convert.ToInt32(columns[6]), Convert.ToInt32(columns[7]));

                            gtEllipse = new GtEllipse(center, axes, angle, imageSize);
                        }
                        return gtEllipse;
                    };
                }
                throw new Exception("Filename not found in the CSV file.");
            }
        }

        /// <summary>
        /// Calculate ellipse's fit score based on fitted and ground truth ellipses' parameters.
        /// </summary>
        /// <param name="fitEllipse">Fitted ellipse's parameters</param>
        /// <param name="gtEllipse">Ground truth ellipse's parameters</param>
        /// <returns>Fitted ellipse's fit score</returns>
        private static float GetEllipseFitScore(Ellipse fitEllipse, GtEllipse gtEllipse)
        {
            Mat fitEllipseImage = DrawEllipse(fitEllipse, gtEllipse.ImageSize);
            Mat gtEllipseImage = DrawEllipse(gtEllipse, gtEllipse.ImageSize);
            Cv2.WaitKey();

            return EvaluateOverlap(gtEllipseImage, fitEllipseImage);
        }

        /// <summary>
        /// Creates a binary image containing an ellipse.
        /// </summary>
        /// <param name="ellipse">Ellipse's parameters</param>
        /// <param name="size">Image's size</param>
        /// <returns>Binary image containing an ellipse</returns>
        private static Mat DrawEllipse(Ellipse ellipse, Size size)
        {
            Mat ellipseImage = Mat.Zeros(size, MatType.CV_8U);
            Point center = new Point((int)Math.Round(ellipse.Center.X), (int)Math.Round(ellipse.Center.Y));
            Size axes = new Size((int)Math.Round(ellipse.Axes.Width), (int)Math.Round(ellipse.Axes.Height));
            double angle = ellipse.Angle;

            Cv2.Ellipse(ellipseImage, center, axes, angle, 0, 360, 1, 1);
            Cv2.Ellipse(ellipseImage, center, axes, angle, 0, 360, 1, Cv2.FILLED);

            return ellipseImage;
        }

        /// <summary>
        /// Calculate overlap of two binary images of ellipses.
        /// </summary>
        /// <param name="gtEllipseImage">Binary image with the ground truth ellipse</param>
        /// <param name="fitEllipseImage">Binary image with the fitted ellipse</param>
        /// <returns>Relative overlap of two binary images of ellipses</returns>
        private static float EvaluateOverlap(Mat gtEllipseImage, Mat fitEllipseImage)
        {
            int fitSum = (int)Cv2.Sum(fitEllipseImage).Val0;
            int gtSum = (int)Cv2.Sum(gtEllipseImage).Val0;

            Mat overlapImage = new Mat();
            Cv2.BitwiseAnd(fitEllipseImage, gtEllipseImage, overlapImage);
            int overlapSum = (int)Cv2.Sum(overlapImage).Val0;

            return overlapSum / (float)Math.Max(fitSum, gtSum);
        }
    }
}
